pycontrails/datalib/gruan.py

```diff
@@ -0,0 +1,280 @@
+                    f"Known sites: {known[product]}"
+                )
+        self.site = site
+
+        if cachestore is self.__marker:
+            cache_root = cache._get_user_cache_dir()
+            cache_dir = f"{cache_root}/gruan"
+            cachestore = cache.DiskCacheStore(cache_dir=cache_dir)
+        self.cachestore = cachestore
+
+        self._ftp: ftplib.FTP | None = None
+
+    def __repr__(self) -> str:
+        return f"GRUAN(product='{self.product}', site='{self.site}')"
+
+    def connect(self) -> ftplib.FTP:
+        """Connect to the GRUAN FTP server."""
+        if self._ftp is None or self._ftp.sock is None:
+            self._ftp = ftplib.FTP(FTP_SERVER)
+            self._ftp.login()
+            return self._ftp
+
+        try:
+            self._ftp.pwd()  # check if connection is still alive
+        except (*ftplib.all_errors, ConnectionError):  # type: ignore[misc]
+            # If we encounter any error, reset the connection and retry
+            self._ftp = None
+            return self.connect()
+        return self._ftp
+
+    @property
+    def base_path_product(self) -> str:
+        """Get the base path for GRUAN data product on the FTP server."""
+        product, version = self.product.rsplit(".")
+        return f"/pub/data/gruan/processing/level2/{product}/version-{version.zfill(3)}"
+
+    @property
+    def base_path_site(self) -> str:
+        """Get the base path for GRUAN data site on the FTP server."""
+        return f"{self.base_path_product}/{self.site}"
+
+    def years(self) -> list[int]:
+        """Get a list of available years for the selected product and site."""
+        ftp = self.connect()
+        ftp.cwd(self.base_path_site)
+        years = ftp.nlst()
+        return [int(year) for year in years]
+
+    def list_files(self, year: int) -> list[str]:
+        """List available files for a given year."""
+        path = f"{self.base_path_site}/{year}"
+
+        ftp = self.connect()
+        try:
+            ftp.cwd(path)
+        except ftplib.error_perm as e:
+            available = self.years()
+            if year not in available:
+                msg = f"No data available for year {year}. Available years are: {available}"
+                raise ValueError(msg) from e
+            raise
+        return ftp.nlst()
+
+    def get(self, filename: str) -> xr.Dataset:
+        """Download a GRUAN dataset by filename."""
+        if self.cachestore is None:
+            return self._get_no_cache(filename)
+        return self._get_with_cache(filename)
+
+    def _get_no_cache(self, filename: str) -> xr.Dataset:
+        t, _ = extract_gruan_time(filename)
+        path = f"{self.base_path_site}/{t.year}/{filename}"
+
+        ftp = self.connect()
+        with tempfile.NamedTemporaryFile(delete=False) as tmp:
+            ftp.retrbinary(f"RETR {path}", tmp.write)
+
+        try:
+            return xr.load_dataset(tmp.name)
+        finally:
+            os.remove(tmp.name)
```

Comment:
```suggestion
        with tempfile.NamedTemporaryFile(delete=True, delete_on_close=False) as tmp:
            ftp.retrbinary(f"RETR {path}", tmp.write)
            tmp.close()
            return xr.load_dataset(tmp.name)
```
I don't think the current code will clean up the temporary file if something goes wrong during `ftp.retrbinary`.

Reply:
Though I don't think the `delete_on_close` kwarg is available before python 3.12, so my solution might be worse...

Reply:
13f8e4ec30d3726a4695400b18a742794638d9bc